src/drivers/mediatek/mt8186/afe-memif.c

```diff
@@ -0,0 +1,719 @@
+}
+
+static int memif_release(struct dma_chan_data *channel)
+{
+	int ret;
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+
+	assert(memif);
+	/* TODO actually handle pause/release properly? */
+	tr_info(&memif_tr, "MEMIF: release(%d)", channel->index);
+
+	if (channel->status != COMP_STATE_PAUSED)
+		return -EINVAL;
+
+	channel->status = COMP_STATE_ACTIVE;
+	ret = afe_memif_set_enable(memif->afe, memif->memif_id, 0);
+	if (ret < 0)
+		return ret;
+#if CONFIG_TEST_SGEN
+	mt8186_afe_sinegen_enable(TEST_SGEN_ID, 48000, 0);
+#endif
+
+	return 0;
+}
+
+static int memif_pause(struct dma_chan_data *channel)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+
+	assert(memif);
+	/* TODO actually handle pause/release properly? */
+	tr_info(&memif_tr, "MEMIF: pause(%d)", channel->index);
+
+	if (channel->status != COMP_STATE_ACTIVE)
+		return -EINVAL;
+
+	channel->status = COMP_STATE_PAUSED;
+
+	/* Disable HW requests */
+	return afe_memif_set_enable(memif->afe, memif->memif_id, 0);
+}
+
+static int memif_stop(struct dma_chan_data *channel)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+
+	assert(memif);
+	tr_info(&memif_tr, "MEMIF: stop(%d)", channel->index);
+	/* Validate state */
+	/* TODO: Should we? */
+	switch (channel->status) {
+	case COMP_STATE_READY:
+	case COMP_STATE_PREPARE:
+		return 0; /* do not try to stop multiple times */
+	case COMP_STATE_PAUSED:
+	case COMP_STATE_ACTIVE:
+		break;
+	default:
+		return -EINVAL;
+	}
+	channel->status = COMP_STATE_READY;
+	/* Disable channel */
+	return afe_memif_set_enable(memif->afe, memif->memif_id, 0);
+}
+
+static int memif_copy(struct dma_chan_data *channel, int bytes, uint32_t flags)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+	struct dma_cb_data next = {
+		.channel = channel,
+		.elem.size = bytes,
+	};
+
+	assert(memif);
+	/* TODO XRUN check, update hw ptr */
+	/* TODO TBD Playback first data check */
+
+	/* update user hwptr */
+	if (memif->direction)
+		memif->wptr = (memif->wptr + bytes) % memif->dma_size;
+	else
+		memif->rptr = (memif->rptr + bytes) % memif->dma_size;
+	tr_dbg(&memif_tr, "memif_copy: wptr:%u, rptr:%u", memif->wptr, memif->rptr);
+
+	notifier_event(channel, NOTIFIER_ID_DMA_COPY, NOTIFIER_TARGET_CORE_LOCAL, &next,
+		       sizeof(next));
+
+	return 0;
+}
+
+static int memif_status(struct dma_chan_data *channel, struct dma_chan_status *status,
+			uint8_t direction)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+	unsigned int hw_ptr;
+
+	assert(memif);
+	status->state = channel->status;
+	status->flags = 0;
+
+	/* update current hw point */
+	hw_ptr = afe_memif_get_cur_position(memif->afe, memif->memif_id);
+	hw_ptr -= memif->dma_base;
+	if (memif->direction)
+		memif->rptr = hw_ptr;
+	else
+		memif->wptr = hw_ptr;
+
+	status->r_pos = memif->rptr + memif->dma_base;
+	status->w_pos = memif->wptr + memif->dma_base;
+	status->timestamp = timer_get_system(timer_get());
+	return 0;
```

Comment:
unless the above comment changes something, this function does not need to return anything.